### DISCUSSION TITLE
stringify-ordering?

### DISCUSSION THREAD

**Discussion by jessekrubin** [samchon/typia#1240] [Q&A]
Is the ordering of the string-ify functions inherited from the type definition? How is the ordering determined?

**Comment by samchon** [comment]
In the object, regular properties (static key named properties) take precedence over dynamic properties.

In the regular properties, optional properties comes earlier than required properties.

  - https://github.com/samchon/typia/blob/master/src/programmers/helpers/StringifyJoinder.ts#L12-L59
  - https://github.com/samchon/typia/blob/master/src/programmers/internal/stringify_regular_properties.ts#L77-L81

**Reply by jessekrubin** [reply]
Thanks! Just to be clear, optional properties come first? Just curious why that is?

It would be kinda cool if using the stringify function you could create a "canonical" format/ordering/sorting

**Comment by samchon** [comment]
Closing as answered by the maintainer.

`stringify` joins regular static-key properties before dynamic properties. Among regular properties, optional properties are emitted before required properties. The implementation references are:

- https://github.com/samchon/typia/blob/master/src/programmers/helpers/StringifyJoinder.ts#L12-L59
- https://github.com/samchon/typia/blob/master/src/programmers/internal/stringify_regular_properties.ts#L77-L81

Do not treat JSON object member order as an interoperability contract outside typia's generated output; if a downstream protocol requires canonical ordering, build that canonical object/string explicitly.
